internal/worker/dbrepl/worker.go

```diff
@@ -0,0 +1,315 @@
+// Copyright 2024 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package dbrepl
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+	"text/tabwriter"
+
+	"github.com/chzyer/readline"
+	"github.com/juju/clock"
+	"github.com/juju/errors"
+	"gopkg.in/tomb.v2"
+
+	"github.com/juju/juju/core/database"
+	coredatabase "github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/logger"
+	"github.com/juju/juju/internal/database/client"
+	"github.com/juju/juju/internal/database/dqlite"
+	"github.com/juju/juju/internal/worker"
+)
+
+// NodeManager creates Dqlite `App` initialisation arguments and options.
+type NodeManager interface {
+	// ClusterServers returns the node information for
+	// Dqlite nodes configured to be in the cluster.
+	ClusterServers(context.Context) ([]dqlite.NodeInfo, error)
+
+	// WithTLSDialer returns a dbApp that can be used to connect to the Dqlite
+	// cluster.
+	WithTLSDialer(ctx context.Context) (client.DialFunc, error)
+}
+
+// DBGetter describes the ability to supply a sql.DB
+// reference for a particular database.
+type DBGetter interface {
+	// GetDB returns a sql.DB reference for the dqlite-backed database that
+	// contains the data for the specified namespace.
+	// A NotFound error is returned if the worker is unaware of the requested DB.
+	GetDB(namespace string) (database.TxnRunner, error)
+}
+
+// WorkerConfig encapsulates the configuration options for the
+// dbaccessor worker.
+type WorkerConfig struct {
+	DBGetter coredatabase.DBGetter
+	Clock    clock.Clock
+	Logger   logger.Logger
+	Stdout   io.Writer
+	Stderr   io.Writer
+	Stdin    io.Reader
+}
+
+// Validate ensures that the config values are valid.
+func (c *WorkerConfig) Validate() error {
+	if c.DBGetter == nil {
+		return errors.NotValidf("missing DBGetter")
+	}
+	if c.Clock == nil {
+		return errors.NotValidf("missing Clock")
+	}
+	if c.Logger == nil {
+		return errors.NotValidf("missing Logger")
+	}
+	return nil
+}
+
+type dbReplWorker struct {
+	cfg  WorkerConfig
+	tomb tomb.Tomb
+
+	dbGetter coredatabase.DBGetter
+}
+
+// NewWorker creates a new dbaccessor worker.
+func NewWorker(cfg WorkerConfig) (*dbReplWorker, error) {
+	var err error
+	if err = cfg.Validate(); err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	w := &dbReplWorker{
+		cfg:      cfg,
+		dbGetter: cfg.DBGetter,
+	}
+
+	w.tomb.Go(w.loop)
+
+	return w, nil
+}
+
+func (w *dbReplWorker) loop() (err error) {
+	ctx, cancel := w.scopedContext()
+	defer cancel()
+
+	history, err := os.CreateTemp("", "juju-repl")
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer history.Close()
```

Comment:
It would be nice if histories were maintained between sessions

Reply:
If not, you should delete the file in a defer as well

Reply:
Yeah, let's delete it for now. I'm not sure of the implications of having queries with potentially sensitive information in them.

We could make it optional to persist it as a flag, but by default it shouldn't.